pydantic/_internal/_core_utils.py

```diff
@@ -498,8 +509,13 @@ def can_be_inlined(s: core_schema.DefinitionReferenceSchema, ref: str) -> bool:
         return True
 
     def inline_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
-        if s['type'] == 'definition-ref':
+        visited = set()
+        while s['type'] == 'definition-ref':
             ref = s['schema_ref']
+            if ref in visited:
+                raise Exception('Circular reference detected')
```

Comment:
Do we need the extra check, if an exception was already thrown in `count_refs`?

Reply:
Good catch, I think checking once will be enough.